### ISSUE TITLE
stream_select() returns 0 when runnnig via apache user on ubuntu

### ISSUE THREAD

**Issue by praveenrchauhan** [phpseclib/phpseclib#1063]
The SSH2 login is failing. When I debugged, I found that `stream_select()` at line no. 1065 in SSH2.php returns `false` when running the script through a browser. However, when I run the same script via terminal, then it works. I'm pretty sure that this isn't a timeout issue. I also tried with RSA key but the same issue happens.

**Comment by bantu** [comment]
The manual says

> On error FALSE is returned and a warning raised (this can happen if the system call is interrupted by an incoming signal).

So is there any warning (possibly logged)?

**Comment by terrafrost** [comment]
I was able to connect via phpseclib via apache user on ubuntu just now without issue..

One possibility: you have selinux installed and httpd_can_network_connect is set to 0. If that's what's going on you can do this to fix it:

```
setsebool -P httpd_can_network_connect=1
```

Also, what version of phpseclib are you using? You're not using the latest 1.0 version or the latest 2.0 version I can tell:

https://github.com/phpseclib/phpseclib/blob/2.0.4/phpseclib/Net/SSH2.php#L1065
https://github.com/phpseclib/phpseclib/blob/1.0.5/phpseclib/Net/SSH2.php#L1065
https://github.com/phpseclib/phpseclib/blob/1.0.3/phpseclib/Net/SSH2.php#L1065
https://github.com/phpseclib/phpseclib/blob/2.0.3/phpseclib/Net/SSH2.php#L1065

That said, the line number does tell me that the issue probably lives in the constructor none-the-less. So one thing that'd help is if you could remove the [error suppression operator](http://php.net/manual/en/language.operators.errorcontrol.php) from `fsockopen` and `stream_select`.

In other words, replace `@fsockopen` with `fsockopen` and `@stream_select` with `stream_select` and tell me if you get any additional messages when running the code that'd be helpful.

Thanks!

**Comment by praveenrchauhan** [comment]
Hi terrafrost,

Thanks for your reply.

- I don't have selinux installed. 
- I am using the copy of `master` branch. `stream_select()` is now called there at line no. 1068 (may be when I had downloaded it last time, it was at L1065 in that version).
https://github.com/phpseclib/phpseclib/blob/master/phpseclib/Net/SSH2.php#L1068
- I had already checked by removing `@` from `fsockopen` and `stream_select` but there was no warning.

Earlier I was using the library without composer but now I have installed the composer and using via it. Now, I can see a warning at that line as follows -

`Warning: stream_select(): You MUST recompile PHP with a larger value of FD_SETSIZE. It is set to 1024, but you have descriptors numbered at least as high as 1041. --enable-fd-setsize=2048 is recommended, but you may want to set it to equal the maximum number of open files supported by your system, in order to avoid seeing this error again at a later date. in <mydir>/phpseclib/Net/SSH2.php on line 1068 `
Hopefully, if the above issue is solved, then it should work.

Could you please let me know, if I should actually proceed with increasing the value for `FD_SETSIZE` or is there any risk involved? 
Or is there an alternative solution?

**Comment by terrafrost** [comment]
I'd say increasing it is a good idea. If you're running into problems with it with PHP's socket_select you're liable to run into problems with it elsewhere (assuming you haven't already had mysterious problems). eg.

http://serverfault.com/questions/300272/getting-error-too-many-open-files-in-system
http://stackoverflow.com/questions/14748499/fatal-error-too-many-open-files

etc.

**Comment by praveenrchauhan** [comment]
Hi terrafrost,

The value `1024` is hard coded into PHP core. I am not sure that after recompiling PHP with an another value for `FD_SETSIZE` will actually work or no but that is something which I would prefer as the last option as  there are couple of functionality running on my machine which I don't want to mess up with.

However, I have figured out that the problem is occurring when I run the script via Apache. I am now running my script through Nginx+HHVM and it is working. As of now, this solution is working for me.

There seems to be some problem with Apache on my setup when running the SSH script.

**Comment by bantu** [comment]
I am going ahead and close this. I would say that it is not directly related to phpseclib anyway. Hope that is fine with you guys.
